File: pt5/1170.compare-strings-by-frequency-of-the-smallest-character.java
import java.util.*;

class Solution {
    public int[] numSmallerByFrequency(String[] queries, String[] words) {
        int[] wordFrequencies = new int[words.length];
        for (int i = 0; i < words.length; i++) {
            wordFrequencies[i] = frequency(words[i]);
        }
        Arrays.sort(wordFrequencies);

        int[] answer = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int queryFrequency = frequency(queries[i]);
            int firstGreater = upperBound(wordFrequencies, queryFrequency);
            answer[i] = wordFrequencies.length - firstGreater;
        }
        return answer;
    }

    private int frequency(String word) {
        char smallest = 'z';
        int count = 0;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            if (ch < smallest) {
                smallest = ch;
                count = 1;
            } else if (ch == smallest) {
                count++;
            }
        }
        return count;
    }

    private int upperBound(int[] values, int target) {
        int left = 0;
        int right = values.length;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (values[mid] <= target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}

File: pt5/1170.compare-strings-by-frequency-of-the-smallest-character.go
package main

import "sort"

func numSmallerByFrequency(queries []string, words []string) []int {
	frequency := func(word string) int {
		smallest := byte('z')
		count := 0
		for i := 0; i < len(word); i++ {
			if word[i] < smallest {
				smallest = word[i]
				count = 1
			} else if word[i] == smallest {
				count++
			}
		}
		return count
	}

	wordFrequencies := make([]int, len(words))
	for i, word := range words {
		wordFrequencies[i] = frequency(word)
	}
	sort.Ints(wordFrequencies)

	answer := make([]int, len(queries))
	for i, query := range queries {
		queryFrequency := frequency(query)
		firstGreater := sort.Search(len(wordFrequencies), func(j int) bool {
			return wordFrequencies[j] > queryFrequency
		})
		answer[i] = len(wordFrequencies) - firstGreater
	}
	return answer
}

